src/core/taylor.rkt

```diff
@@ -70,90 +74,86 @@
+  (define taylor-approxs (make-vector (vector-length nodes)))
+  (for ([expr (in-vector nodes)] [n (in-naturals)])
+    (define approx
+      (match expr
+        [(? (curry equal? var)) (taylor-exact 0 1)]
+        [(? number?) (taylor-exact expr)]
+        [(? variable?) (taylor-exact expr)]
+        [`(,const) (taylor-exact expr)]
+        [`(+ ,args ...) (apply taylor-add (map (curry vector-ref taylor-approxs) args))]
+        [`(neg ,arg) (taylor-negate ((curry vector-ref taylor-approxs) arg))]
+        [`(- ,arg1 ,arg2)
+         (taylor-add (vector-ref taylor-approxs arg1)
+                     (taylor-negate (vector-ref taylor-approxs arg2)))]
+        [`(* ,left ,right)
+         (taylor-mult (vector-ref taylor-approxs left) (vector-ref taylor-approxs right))]
+        [`(/ ,num ,den)
+         #:when (equal? (vector-ref nodes num) 1)
+         (taylor-invert (vector-ref taylor-approxs den))]
+        [`(/ ,num ,den)
+         (taylor-quotient (vector-ref taylor-approxs num) (vector-ref taylor-approxs den))]
+        [`(sqrt ,arg) (taylor-sqrt var (vector-ref taylor-approxs arg))]
+        [`(cbrt ,arg) (taylor-cbrt var (vector-ref taylor-approxs arg))]
+        [`(exp ,arg) (taylor-exp (vector-ref taylor-approxs arg) `(exp ,(get-expr nodes arg)))]
+        [`(sin ,arg) (taylor-sin (vector-ref taylor-approxs arg) `(sin ,(get-expr nodes arg)))]
+        [`(cos ,arg) (taylor-cos (vector-ref taylor-approxs arg) `(cos ,(get-expr nodes arg)))]
+        [`(tan ,arg)
+         (taylor-quotient (taylor-sin (vector-ref taylor-approxs arg) `(sin ,(get-expr nodes arg)))
+                          (taylor-cos (vector-ref taylor-approxs arg) `(cos ,(get-expr nodes arg))))]
+        [`(log ,arg) (taylor-log var (vector-ref taylor-approxs arg))]
+        [`(pow ,base ,power)
+         #:when (exact-integer? (vector-ref nodes power))
+         (taylor-pow (normalize-series (vector-ref taylor-approxs base)) (vector-ref nodes power))]
+        [`(pow ,base ,power)
+         #:when (equal? (vector-ref nodes power) 1/2)
+         (taylor-sqrt var (vector-ref taylor-approxs base))]
+        [`(pow ,base ,power)
+         #:when (equal? (vector-ref nodes power) 1/3)
+         (taylor-cbrt var (vector-ref taylor-approxs base))]
+        [`(pow ,base ,power)
+         #:when (equal? (vector-ref nodes power) 2/3)
+         (define tx (vector-ref taylor-approxs base))
+         (taylor-cbrt var (taylor-mult tx tx))]
+        [`(pow ,base ,power) ; `(exp (* ,power (log ,base)))
+         (taylor-exp (taylor-mult (vector-ref taylor-approxs power)
+                                  (taylor-log var (vector-ref taylor-approxs base)))
+                     `(exp (* ,(get-expr nodes power) (log ,(get-expr nodes base)))))]
+        [`(sinh ,arg)
+         (define exparg (taylor-exp (vector-ref taylor-approxs arg) `(exp ,(get-expr nodes arg))))
+         (taylor-mult (taylor-exact 1/2) (taylor-add exparg (taylor-negate (taylor-invert exparg))))]
+        [`(cosh ,arg)
```

Comment:
Some of these cases like cosh are just 1) expanded to definitions, which are 2) manually deduplicated. I wonder if that can be simplified now that batches deduplicate anyway…

Reply:
I can eliminate it at `progs->batch` stage by introducing some specific taylor-rewriting rules when, say, flag `taylor` is true.
One of the rules can be: parse `(cosh arg)` argument as `(cosh (exp arg))`, since the rule for `cosh` doesn't interact with `arg` at all, only with `(exp arg)`
I see in profiling that `taylor-exp` takes a lot. The changes to `progs->batch` may be helpful

Reply:
@pavpanchekha 
I can loop through `expr` in this manner either in `progs->batch` or simply in `taylor` function:
```
(for/list ([e (in-list expr)])
  (match e
    [(list 'cosh args ...) (list 'cosh (cons 'exp args))]
    [...]
    [_ e]))
```
This may eliminate any duplications,
but from another point of view it will split `taylor` code into possibly two files and the reading will be harder

Reply:
Changed, all the rewritings moved to `progs->batch`. 
If the code stays as it is now, we do not need `tanh`, `cosh` and etc matches at `taylor`, because they are done in `progs->batch`.

Reply:
Uhh, let's delete them, it's a good check to make sure that code isn't being exercised.